### PR TITLE
refactor(pubsub): make helper linking functions so they can be shared

### PR DESCRIPTION
**Pull request by alevenberg**
Closes https://github.com/googleapis/google-cloud-cpp/issues/13265

- Makes `CreateLinks` and `MaybeAddLinkAttributes` more generic so they can be re-used in other places

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13371)

### CODE REVIEW

File: google/cloud/pubsub/internal/tracing_pull_ack_handler.cc [hunk 1]
```diff
@@ -38,36 +38,39 @@ using Attributes =
 
 namespace {
 
-auto CreateLinks() {
+/// Create a list of links with @p span context if compiled with open telemetery
+/// ABI 2.0.
+auto CreateLinks(
+    std::shared_ptr<opentelemetry::trace::SpanContext> const& span_context) {
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_pull_ack_handler.cc:44]: passing a `shared_ptr<>` as a `const&` is a code smell.  You probably want:

```suggestion
auto CreateLinks(opentelemetry::trace::SpanContext const& span_context) {
```

File: google/cloud/pubsub/internal/tracing_pull_ack_handler.cc [hunk 2]
```diff
@@ -38,36 +38,39 @@ using Attributes =
...
 }
 
+/// Adds two link attributes to the @p current span for the trace id and span id
+/// from @span.
 void MaybeAddLinkAttributes(
-    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> const& span) {
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> const&
+        current_span,
+    std::shared_ptr<opentelemetry::trace::SpanContext> const& span_context,
+    std::string const& span_name) {
```
Comment (coryan) [inline google/cloud/pubsub/internal/tracing_pull_ack_handler.cc:62]: I should have noticed these before, but same comment:

```suggestion
    opentelemetry::trace::Span const& current_span,
    opentelemetry::trace::SpanContext const& span_context,
    std::string const& span_name) {
```

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/pubsub/internal/tracing_pull_ack_handler.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_pull_ack_handler.cc:62]: Done

File: google/cloud/pubsub/internal/tracing_pull_ack_handler.cc (hunk 1 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/tracing_pull_ack_handler.cc:44]: Done

**Review by coryan — APPROVED** [review]

**Comment by alevenberg** [comment]
```Step #3: 516/543 Test #321: storage_internal_grpc_stub_failures_test ..................................***Exception: SegFault  0.79 sec
Step #3: Running main() from gmock_main.cc
Step #3: Note: Randomizing tests' orders with a seed of 37174 .
Step #3: [==========] Running 98 tests from 2 test suites.
Step #3: [----------] Global test environment set-up.
Step #3: [----------] 49 tests from HybridClientFailures/GrpcClientFailuresTest
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.GetNativeBucketIamPolicy/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.GetNativeBucketIamPolicy/0 (32 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.UpdateBucket/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.UpdateBucket/0 (30 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.SetNativeBucketIamPolicy/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.SetNativeBucketIamPolicy/0 (29 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.SignBlob/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.SignBlob/0 (35 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.ListObjects/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.ListObjects/0 (32 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.ListNotifications/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.ListNotifications/0 (32 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.GetBucketAcl/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.GetBucketAcl/0 (36 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.RewriteObject/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.RewriteObject/0 (40 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.ComposeObject/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.ComposeObject/0 (34 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.InsertObjectMediaMultipart/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.InsertObjectMediaMultipart/0 (34 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.GetServiceAccount/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.GetServiceAccount/0 (30 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.UpdateObjectAcl/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.UpdateObjectAcl/0 (26 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.ListObjectAcl/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.ListObjectAcl/0 (26 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.DeleteBucket/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.DeleteBucket/0 (38 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.GetBucketMetadata/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.GetBucketMetadata/0 (35 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.PatchObject/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.PatchObject/0 (44 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.DeleteDefaultObjectAcl/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.DeleteDefaultObjectAcl/0 (45 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.CreateDefaultObjectAcl/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.CreateDefaultObjectAcl/0 (41 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.TestBucketIamPermissions/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.TestBucketIamPermissions/0 (53 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.CreateBucket/0
Step #3: [       OK ] HybridClientFailures/GrpcClientFailuresTest.CreateBucket/0 (40 ms)
Step #3: [ RUN      ] HybridClientFailures/GrpcClientFailuresTest.CopyObject/0
Step #3: 
Step #3: 517/543 Test #591: bigtable_table_apply_test .................................................   Passed    0.46 sec
Step #3: 518/543 Test #701: spanner_internal_transaction_impl_test ....................................   Passed    0.21 sec
Step #3: 519/543 Test #589: bigtable_rpc_retry_policy_test ............................................   Passed    0.49 sec
Step #3: 520/543 Test #477: pubsub_internal_subscription_session_test .................................   Passed    0.76 sec
Step #3: 521/543 Test #609: bigtable_bigtable_benchmark_test ..........................................   Passed    0.51 sec
Step #3: 522/543 Test #675: spanner_bytes_test ........................................................   Passed    0.52 sec
Step #3: 523/543 Test #665: pubsublite_internal_resumable_async_streaming_read_write_rpc_test .........   Passed    0.58 sec
Step #3: 524/543 Test #677: spanner_client_test .......................................................   Passed    0.68 sec
Step #3: 525/543 Test #293: storage_internal_async_partial_upload_test ................................   Passed    1.27 sec
Step #3: 526/543 Test #687: spanner_internal_connection_impl_test .....................................   Passed    0.89 sec
Step #3: 527/543 Test  #72: common_internal_async_connection_ready_test ...............................   Passed    2.02 sec
Step #3: 528/543 Test #190: storage_client_bucket_test ................................................   Passed    1.91 sec
Step #3: 529/543 Test #724: spanner_bytes_benchmark ...................................................   Passed    1.49 sec
Step #3: 530/543 Test #200: storage_client_test .......................................................   Passed    2.53 sec
Step #3: 531/543 Test #727: spanner_row_benchmark .....................................................   Passed    1.63 sec
Step #3: 532/543 Test #658: pubsublite_internal_alarm_registry_impl_test ..............................   Passed    2.01 sec
Step #3: 533/543 Test #662: pubsublite_internal_multipartition_publisher_test .........................   Passed    2.02 sec
Step #3: 534/543 Test #319: storage_internal_grpc_split_write_object_data_test ........................   Passed    2.47 sec
Step #3: 535/543 Test #476: pubsub_internal_subscription_message_queue_test ...........................   Passed    2.38 sec
Step #3: 536/543 Test #309: storage_internal_grpc_make_cord_test ......................................   Passed    2.57 sec
Step #3: 537/543 Test #284: storage_internal_crc32c_benchmark .........................................   Passed    3.52 sec
Step #3: 538/543 Test #725: spanner_internal_merge_chunk_benchmark ....................................   Passed    3.60 sec
Step #3: 539/543 Test  #68: common_completion_queue_test ..............................................   Passed    5.00 sec
Step #3: 540/543 Test  #66: common_options_benchmark ..................................................   Passed    5.91 sec
Step #3: 541/543 Test #726: spanner_numeric_benchmark .................................................   Passed    7.42 sec
Step #3: 542/543 Test #116: common_completion_queue_benchmark .........................................   Passed   12.92 sec
Step #3: 543/543 Test #165: common_internal_curl_handle_factory_benchmark .............................   Passed   19.23 sec``` 

Looks unrelated.

**Comment by alevenberg** [comment]
Failed build here: https://pantheon.corp.google.com/cloud-build/builds;region=us-east1/c14f5035-22e9-433d-98e8-6db01b59e49d;step=3?e=-13802955&mods=logs_tg_prod&project=cloud-cpp-testing-resources

**Pull request merged**